### ISSUE TITLE
Image.toByteData()

### ISSUE THREAD

**Issue by collinjackson** [flutter/flutter#11648]
In #6774 @Hixie proposed a `toByteData` method for the `Image` class. This would allow reading the raw pixels out of an image.

@abarth added support for `Picture.toImage` which closed #6774 but `Image` still doesn't have a `toByteData` method. @sethladd asked about this on [Stack Overflow](https://stackoverflow.com/questions/45680088/how-to-save-a-flutter-canvas-as-a-bitmap-image) so I created this issue to track `toByteData` implementation progress. I think it's probably a blocker for #1433.

**Comment by chinmaygarde** [comment]
For an image backed by an OpenGL texture, we discard the host side copy if the image. So there is no byte data available to return. We could either keep these copies around or perform an expensive draw call and read pixels after to populate the byte buffer on demand.

**Comment by chinmaygarde** [comment]
We can also create a software rendering context and render that image into the same. In any case though, such an operation will be extremely expensive.

**Comment by sethladd** [comment]
Can it be an async call?

**Comment by collinjackson** [comment]
It's probably fine for it to be slow if this is only used for #1433.

**Comment by abarth** [comment]
Needs to be async.

**Comment by sethladd** [comment]
> Needs to be async.

👍

**Comment by tvolkert** [comment]
Within Google, several customers use screen diffing tests with great success.  They're currently required to do this in `flutter_driver` tests, but if we had `Image.toByteData`, that would unlock our ability to provide screen diffing tests in `flutter_tester` tests as well.

This would be a big win on the Flutter team, since there have been several times that we've broken stuff that only image diffing would have caught (e.g. https://github.com/flutter/flutter/issues/12820, which was only caught by a Google customer's internal tests).

@eseidel and I were talking about how we can catch problems earlier in the pipeline (e.g. strive to never learn of a framework breakage only via a customer test failure).  This is one way.

**Comment by eseidelGoogle** [comment]
My understanding is the screen diffing test could be orthogonal to this question.  Since those could be implemented with a custom observatory-protocol service which knew how to serve screenshots rather than this more general purpose "get the drawn bytes from a Picture".  But yes, screen diffing could also be implementing on top of this bug, were it to be resolved first.

**Comment by Hixie** [comment]
I propose we add two methods:

```dart
  Future<Uint8List> Image.toByteData();
  Future<Uint8List> Picture.toByteData();
```

They would both return raw pixel data in RGBA form, row-primary. The Image one would be slower since Images are GPU-backed. The Picture one would be faster since it could create a CPU-side image first.

cc @chinmaygarde

**Comment by lejard-h** [comment]
Do you have a roadmap on this ? I need to create a signature area for a client, I can easily draw on screen, but this is blocking to get the image of the signature.

**Comment by sethladd** [comment]
Hi @lejard-h ! Thanks for reaching out. Curious, would you be willing to share more about the app that you're building? If you want to share in private, feel free to email me :)

**Comment by johnpryan** [comment]
@sethladd we are also wanting this at AppTree for handling user signatures:

We have a [branch of signature-pad-dart](https://github.com/johnpryan/signature-pad-dart/tree/flutter) that works with flutter but it can't currently be used to save the signature as an image.

our use case: get a signature from the user as a bitmap or other image type and upload to the server

**Comment by sroddy** [comment]
our use case: save a drawing to the camera roll, send it to the share widget

almost a road blocker for at least 3/4 of our apps

**Comment by sethladd** [comment]
@johnpryan I've marked this as critical. Is it a blocker?

**Comment by gedw99** [comment]
I have similar need. Screenshot the widget tree and share to printer. This allows printing a screen with a customer's signature on a form type of use case.

Screen differencing to catch regressions is something we use currently with non flutter GUI and since new projects are being done in flutter I would be happy as Larry if we could detect differences and export those within the app. Not via observatory.

**Comment by johnpryan** [comment]
@sethladd Thanks - this is not necessarily a blocker. If this issue is a large effort a plugin could be built instead.

**Comment by kentcb** [comment]
Arrived here looking for the same thing (to support a signature pad). Also blocked by this.

**Comment by lejard-h** [comment]
The only workaround I could think is tu use a webview that serve a [signature_pad](https://github.com/szimek/signature_pad) and use the [javascript](https://github.com/dart-flitter/flutter_webview_plugin/blob/master/lib/src/base.dart#L101) api to manipulate the view and get the signature as an image.

**Comment by gedw99** [comment]
Webview ist a good solution for now for me.
You want a nice big single view to scrawl on anyway and there is code that
does this out there now.

**Comment by johnpryan** [comment]
@lejard-h no javascript allowed! https://github.com/apptreesoftware/signature-pad-dart

**Comment by nsreenath** [comment]
@Hixie is there any work going on related to this ?

**Comment by Hixie** [comment]
This is not in our currently priority list, but it's high on our list for things to get to in a few months since it's critical for a variety of features we care about (including, in particular, improving our testing). We would accept patches for this feature if anyone is interested in developing it sooner.

**Comment by terrywarwar** [comment]
+3 apps

**Comment by alexshikov** [comment]
Has anybody found a workaround?

**Comment by e-simpson** [comment]
@alexshikov This was fixed in PR [#4762](https://github.com/flutter/engine/pull/4762) mentioned above. You can now use Image.toByteData(). If you do not see this functionality update your Flutter engine.

**Comment by johnpryan** [comment]
@alexshikov See [this StackOverflow answer](https://stackoverflow.com/a/52524417/5859156)
